src/engines/hugin-inference-engine.ts

```diff
@@ -0,0 +1,226 @@
+  // time that there are changes to the evidence, or CPT for the
+  // network's variables.
+  private _potentials: ICliquePotentials = {}
+  private _marginals: INetworkResult = {}
+
+  // Internal store of the current evidence for the engine.
+  // This can only be mutated by the provided functions.
+  private _evidence: ICombinations;
+
+  // The constructor instanciates all the internally cached information
+  // about the bayes network structure and initiallzes the evidence to
+  // void.
+  constructor (network: INetwork) {
+    this._network = clone(network) // note that cloning here is required to prevent external mutation of the network.
+    this._evidence = {}
+    const { cliques, sepSets, junctionTree } = createCliques(network)
+    this._connectedComponents = getConnectedComponents(junctionTree)
+    this._cliques = cliques
+    this._sepSets = sepSets
+    this._junctionTree = junctionTree
+  }
+
+  // A helper method for resetting the cached potentials and marginals.
+  private resetCache = () => {
+    this._potentials = {}
+    this._marginals = {}
+  }
+
+  // Implementation of various member functions
+  hasVariable = (name: string) => this._network[name] != null;
+  getVariables = () => Object.keys(this._network);
+
+  hasLevel = (name: string, level: string) => {
+    if (!this._network[name]) return false
+    const found = this._network[name].states.filter(x => x === level)
+    return found !== []
+  }
+
+  getLevels = (name: string) => {
+    if (!this._network[name]) return []
+    return [...this._network[name].states]
+  }
+
+  hasParent = (name: string, parent: string) => {
+    if (!this._network[name]) return false
+    const found = this._network[name].parents.filter(x => x === parent)
+    return found !== []
+  }
+
+  getParents = (name: string) => {
+    if (!this._network[name]) return []
+    return [...this._network[name].parents]
+  }
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // result is cloned prior to returning it.
+  getDistribution = (name: string) => clone(this._network[name].cpt)
+
+  // NOTE: in order to prevent external mutation of the CPT, the
+  // input is cloned prior to assigining to the internally cached value.
+  setDistribution = (name: string, cpt: ICptWithParents | ICptWithoutParents) => {
+    const node: INode = this._network[name]
+    const expectedLevels = node.states
+    const expectedParents = node.parents
+
+    // Perform some sanity checks.
+    const err = (reason: string) => {
+      throw new Error(`Cannot set the distribution for ${name}.  ${reason}.`)
+    }
+    if (!node) err('The variable does not exist in the network')
+
+    const observedLevels = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].then)
+        : []
+      : Object.keys(cpt).sort()
+
+    const observedParents = Array.isArray(cpt)
+      ? cpt.length > 0
+        ? Object.keys(cpt[0].when)
+        : []
+      : []
+
```

Comment:
These nested ternaries it's kind of bad to understand.... my suggestion would be to convert both into private methods:

```ts
private getObservedLevels(cpt) {
  if (Array.isArray(cpt)) {
    if (!isEmpty(cpt)) {
      return Object.keys(cpt[0].when)
    }
  }

  return []
}
```